app/src/main/kotlin/com/x8bit/bitwarden/data/auth/manager/KdfManagerImpl.kt

```diff
@@ -0,0 +1,105 @@
+        if (!featureFlagManager.getFeatureFlag(FlagKey.ForceUpdateKdfSettings)) {
+            return false
+        }
+
+        val account = authDiskSource
+            .userState
+            ?.accounts
+            ?.get(activeUserId)
+            ?: return false
+
+        if (account.profile.userDecryptionOptions != null &&
+            !account.profile.userDecryptionOptions.hasMasterPassword
+        ) {
+            return false
+        }
+
+        return account.profile.kdfType == KdfTypeJson.PBKDF2_SHA256 &&
+            account.profile.kdfIterations != null &&
+            account.profile.kdfIterations < DEFAULT_PBKDF2_ITERATIONS
+    }
+
+    override suspend fun updateKdfToMinimumsIfNeeded(password: String): UpdateKdfMinimumsResult {
+        val userId = activeUserId ?: return UpdateKdfMinimumsResult.ActiveAccountNotFound
+
+        // Check if needs update kdf
+        if (!needsKdfUpdateToMinimums()) {
+            return UpdateKdfMinimumsResult.Success
+        }
+
+        val defaultKdf = Kdf.Pbkdf2(iterations = DEFAULT_PBKDF2_ITERATIONS.toUInt())
+        // Generate updated KDF data
+        val updateKdfResponse = vaultSdkSource
+            .makeUpdateKdf(
+                userId = userId,
+                password = password,
+                kdf = defaultKdf,
+            )
+            .getOrElse { error ->
+                return UpdateKdfMinimumsResult.Error(error = error)
+            }
+
+        val authData = updateKdfResponse.masterPasswordAuthenticationData
+        val oldAuthData = updateKdfResponse.oldMasterPasswordAuthenticationData
+        val unlockData = updateKdfResponse.masterPasswordUnlockData
+        // Send update to server
+        val updateKdfRequest = UpdateKdfJsonRequest(
+            authenticationData = MasterPasswordAuthenticationDataJson(
+                kdf = authData.kdf.toKdfRequestModel(),
+                masterPasswordAuthenticationHash = authData.masterPasswordAuthenticationHash,
+                salt = authData.salt,
+            ),
+            key = unlockData.masterKeyWrappedUserKey,
+            masterPasswordHash = oldAuthData.masterPasswordAuthenticationHash,
+            newMasterPasswordHash = authData.masterPasswordAuthenticationHash,
+            unlockData = MasterPasswordUnlockDataJson(
+                kdf = unlockData.kdf.toKdfRequestModel(),
+                masterKeyWrappedUserKey = unlockData.masterKeyWrappedUserKey,
+                salt = unlockData.salt,
+            ),
+        )
+
+        accountsService
+            .updateKdf(body = updateKdfRequest)
+            .getOrElse { error ->
+                return UpdateKdfMinimumsResult.Error(error = error)
+            }
+
+        val currentUserState = authDiskSource.userState
+        // Update profile with new KDF parameters
+        authDiskSource.userState = currentUserState?.toUserStateJsonKdfUpdatedMinimums()
+
+        return UpdateKdfMinimumsResult.Success
```

Comment:
🎨 Could use `fold()` here.

```suggestion
        return accountsService
            .updateKdf(body = updateKdfRequest)
            .fold(
                onSuccess = {
                    val currentUserState = authDiskSource.userState
                    // Update profile with new KDF parameters
                    authDiskSource.userState = currentUserState?.toUserStateJsonKdfUpdatedMinimums()
                    UpdateKdfMinimumsResult.Success
                },
                onFailure = { UpdateKdfMinimumsResult.Error(error = it) },
            )
```